### ISSUE TITLE
Make the Sound Blaster code consistent with other audio devices

### ISSUE THREAD

**Issue by kcgen** [dosbox-staging/dosbox-staging#1789]
- [ ] Refactor it into a class
- [ ] Switch to the `MaybeRenderFrame()` callback approach
- [ ] Generate floating-point frames

When all audio devices support this, the frame handling logic can be generalized in the mixer instead of existing in each audio device.

**Comment by johnnovak** [comment]
> Maybe related:
> 
>     * [[post-0.82.0] Standardize the Sound Blaster against other audio devices #3875](https://github.com/dosbox-staging/dosbox-staging/pull/3875)

@Torinde, please explain it to me, if you don't know if it's related or not, why did you make this comment? Boredom?

**Comment by Grounded0** [comment]
I'm quitting all issue tracker duties so its good we still have someone sleuthing these old tickets since no one else does.

Looks like its close enough in its intent to potentially close this ticket when its merged.

**Comment by interloper98** [comment]
> Looks like its close enough in its intent to potentially close this ticket when its merged.

My PR definitely gets the third item.

For the second item, the PR always renders per-frame at exactly the playback 
 rate (DMA and DAC); so I think it hit that point, just without the 'Maybe' aspect.

As for the first point (turning SB into a a class): I definitely saw this too - but I haven't touched it in my PR. Right now there's a global SB struct and a separate config class; it would be nice to combine all these and use member functions instead of static functions. Maybe I'll take that on later.

Thanks for connecting these points @Torinde. I'll mention this in the PR wording.

**Comment by johnnovak** [comment]
> I'm quitting all issue tracker duties so its good we still have someone sleuthing these old tickets since no one else does.

Except for me and @Burrito78 a couole of times a year 😉 

> Right now there's a global SB struct and a separate config class; it would be nice to combine all these and use member functions instead of static functions. Maybe I'll take that on later.

Yeah, could be done, but we don't really need individual tickets for these refactor jobs IMO. This was a reminder for kcgen I think.

Then we have a ticket for the float conversion of all audio devices, plus yeah there's @interloper98's PR so I think this one can be closed.

> I was fairly certain it's related (c'mon, just by reading the titles it seems related). And by looking at the PR to me it seemed some things from here are accomplished in the PR. But I'm not confident to claim all that as fact, so I just wanted to notify you guys, so that you can.

Ok, thanks 😄 You just meddle a lot in dev related tickets for a non-developer, so I'm a bit wary when I see your comments these days 😆

**Comment by Grounded0** [comment]
@Torinde 

You're not the only person getting berated for doing the right thing lately:

- #1821

**Comment by Grounded0** [comment]
Turned top post into a checklist.

**Comment by johnnovak** [comment]
@interloper98 Your current PR addresses most (all?) points of this issue ticket, right?

**Comment by interloper98** [comment]
> @interloper98 Your current PR addresses most (all?) points of this issue ticket, right?

Yes. Also tested that prior commits compile and run, and regression tested it against the older SB tickets. Nothing left on my side.

**Comment by johnnovak** [comment]
@weirddan455 You've done most (all?) of the bullet points from the top description, right?

**Comment by weirddan455** [comment]
I haven't done any of those things. @interloper98 did some refactoring work on SoundBlaster and I think it's now converting to floating-point in the SoundBlaster code.

I don't know what "Switch to the MaybeRenderFrame() callback approach" means or how it would be helpful. We have both per-frame and per-tick callbacks.

**Comment by johnnovak** [comment]
> I haven't done any of those things. [@interloper98](https://github.com/interloper98) did some refactoring work on SoundBlaster and I think it's now converting to floating-point in the SoundBlaster code.
> 
> I don't know what "Switch to the MaybeRenderFrame() callback approach" means or how it would be helpful. We have both per-frame and per-tick callbacks.

Cheers, it's a 2022 ticket, so I'll just close it after all as kcgen is no longer with us.

The floating point stuff will be addressed by me at some point, I have a ticket for that.

For the rest, probably those plans are out of date, yeah.
